frame/balances/src/lib.rs

```diff
@@ -1148,59 +1109,74 @@ impl<T: Trait<I>, I: Instance> Currency<T::AccountId> for Module<T, I> where
+				.ok_or(Error::<T, I>::InsufficientBalance)?;
 
-		// bail if not yet created and this operation wouldn't be enough to create it.
-		if value < ed && account.total().is_zero() { return Self::PositiveImbalance::zero() }
+			// bail if we need to keep the account alive and this would kill it.
+			let ed = T::ExistentialDeposit::get();
+			let would_kill = new_free_account < ed && account.free >= ed;
+			ensure!(liveness == AllowDeath || !would_kill, Error::<T, I>::KeepAlive);
 
-		// defensive only: overflow should never happen, however in case it does, then this
-		// operation is a no-op.
-		account.free = match account.free.checked_add(&value) {
-			Some(f) => f,
-			None => return Self::PositiveImbalance::zero(),
-		};
+			Self::ensure_can_withdraw(who, value, reasons, new_free_account)?;
 
-		Self::set_account(who, &account, &old_account);
+			account.free = new_free_account;
 
-		PositiveImbalance::new(value)
+			Ok(NegativeImbalance::new(value))
+		})
 	}
 
 	/// Force the new free balance of a target account `who` to some new value `balance`.
-	fn make_free_balance_be(who: &T::AccountId, value: Self::Balance) -> (
-		SignedImbalance<Self::Balance, Self::PositiveImbalance>,
-		UpdateBalanceOutcome
-	) {
-		let old_account = Self::account(who);
-		let mut account = old_account.clone();
-
-		if value < T::ExistentialDeposit::get() && account.free.is_zero() {
+	fn make_free_balance_be(who: &T::AccountId, value: Self::Balance)
+		-> SignedImbalance<Self::Balance, Self::PositiveImbalance>
+	{
+		Self::try_mutate_account(who, |account|
+			-> Result<SignedImbalance<Self::Balance, Self::PositiveImbalance>, ()>
+		{
+			let ed = T::ExistentialDeposit::get();
 			// If we're attempting to set an existing account to less than ED, then
 			// bypass the entire operation. It's a no-op if you follow it through, but
 			// since this is an instance where we might account for a negative imbalance
 			// (in the dust cleaner of set_account) before we account for its actual
 			// equal and opposite cause (returned as an Imbalance), then in the
 			// instance that there's no other accounts on the system at all, we might
 			// underflow the issuance and our arithmetic will be off.
-			return (
-				SignedImbalance::Positive(Self::PositiveImbalance::zero()),
-				UpdateBalanceOutcome::AccountKilled,
-			)
-		}
-		let imbalance = if account.free <= value {
-			SignedImbalance::Positive(PositiveImbalance::new(value - account.free))
-		} else {
-			SignedImbalance::Negative(NegativeImbalance::new(account.free - value))
-		};
-		account.free = value;
-
-		// If the balance is too low, then the account is reaped.
-		// Free balance can never be less than ED. If that happens, it gets reduced to zero
-		// and the account information relevant to this subsystem is deleted (i.e. the
-		// account is reaped).
-		let outcome = Self::set_account(who, &account, &old_account);
-		(imbalance, outcome)
+			ensure!(value >= ed || !account.free.is_zero(), ());
```

Comment:
This logic is not right anymore I think.

We should check that value + existing reserved value >= ed.

Also, why does this fail now if `account.free.is_zero()`. We can't call this function on account with some free balance already?